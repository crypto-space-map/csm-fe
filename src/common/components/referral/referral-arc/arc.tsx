import React from 'react';

const getCircleLength = (radius: number) => Math.PI * radius * 2;

type ReferralArcProps = React.SVGAttributes<HTMLOrSVGElement> & {
  percents: number;
  radius?: number;
};

export const ReferralArc = ({ percents, radius = 70, ...rest }: ReferralArcProps) => {
  const dashArray = getCircleLength(radius);
  const strokeDashOffset = dashArray - dashArray * (percents / 2 / 100);
  return (
    <svg
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
      viewBox="0 0 180 180"
      {...rest}>
      <g transform="rotate(-180 90 100)">
        <circle
          vectorEffect="non-scaling-stroke"
          cx="50%"
          cy="60%"
          r={radius}
          stroke="#A9F89D"
          strokeWidth="12"
          strokeLinecap="round"
          fill="none"
          strokeDasharray={dashArray}
          strokeDashoffset={strokeDashOffset}
        />
      </g>
    </svg>
  );
};

ReferralArc.defaultProps = {
  radius: 70,
};
