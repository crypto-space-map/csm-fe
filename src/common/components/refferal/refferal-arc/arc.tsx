import { CanvasHTMLAttributes, useCallback, useEffect, useMemo, useRef, useState } from 'react';

import { COLOR_PALLETTE } from 'global/pallette';

export type RefferalArcProps = CanvasHTMLAttributes<HTMLCanvasElement> & {
  percents: number;
  radius?: number;
};

export const RefferalArc = ({ percents, radius = 70, ...rest }: RefferalArcProps) => {
  const ref = useRef<HTMLCanvasElement>(null);
  const backgroundRef = useRef<HTMLCanvasElement>(null);
  const [ctx, setCtx] = useState<CanvasRenderingContext2D | null | undefined>(null);
  const [backgroundCtx, setBackgroundCtx] = useState<CanvasRenderingContext2D | null | undefined>(null);

  useEffect(() => {
    setCtx(ref.current?.getContext('2d'));
    setBackgroundCtx(backgroundRef.current?.getContext('2d'));
  }, []);

  /**
   *  Current usage of calculated value is 1 = whole number(integer)
   *
   *  (percents / 100) - get a part as a percentage of an integer
   *
   *  output F.ex percentage = 48
   *
   *  second part of func = 48/100 = 0.48
   *  output calculatedValue = 1 - 0.48 = 0.52 ( this value will be subtracted like endAngle)
   * */

  const INTEGER = 1;
  const DIVIDER = 100;

  const calculatedValue = INTEGER - percents / DIVIDER;
  const startAngle = Math.PI / -1;
  const endAngle = -calculatedValue * Math.PI;
  const lineWidth = 12;

  const canvas = useMemo(
    () => ({ height: radius + lineWidth * 2, width: radius * 2 + lineWidth * 2 }),
    [radius]
  );

  const ctxRender = useCallback(
    (context: CanvasRenderingContext2D, fullWidth = null) => {
      context.beginPath();
      context.lineWidth = lineWidth;
      context.lineCap = 'round';
      context.strokeStyle = fullWidth ? COLOR_PALLETTE.MAIN_LAYOUT : COLOR_PALLETTE.MAIN_GREEN;
      context.arc(radius + lineWidth, radius + lineWidth, radius, startAngle, fullWidth || endAngle, false);
      context.stroke();
    },
    [radius, startAngle, endAngle]
  );

  useEffect(() => {
    if (ctx && backgroundCtx) {
      ctxRender(ctx);
      ctxRender(backgroundCtx, '-0');
    }
  }, [backgroundCtx, ctx, ctxRender, endAngle, radius, startAngle]);

  return (
    <>
      <canvas {...canvas} ref={backgroundRef} {...rest} /> <canvas {...canvas} ref={ref} {...rest} />
    </>
  );
};
