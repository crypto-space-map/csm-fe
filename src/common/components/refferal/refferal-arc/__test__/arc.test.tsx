import { render } from '@testing-library/react';

import { RefferalArc } from '../arc';

describe('<RefferalArc />', () => {
  it('should match snapshot', () => {
    const { container } = render(<RefferalArc percents={5} radius={70} />);
    expect(container).toMatchSnapshot();
  });
  it('should be defined', () => {
    const { container } = render(<RefferalArc percents={100} radius={80} />);
    expect(container).toBeDefined();
  });
});
