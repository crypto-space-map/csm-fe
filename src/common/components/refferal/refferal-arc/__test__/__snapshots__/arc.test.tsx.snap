// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RefferalArc /> should match snapshot 1`] = `
<div>
  <canvas
    height="94"
    width="164"
  />
</div>
`;
